feat: Add heartbeat mechanism for failure discovery

vpapavas opened the pull request:
### Description 
Added `HeartbeatHandler` and two resources `HeartbeatResource` and `ClusterStatusResource`

### Testing done 
Check of tests for `HeartbeatHandler` and `HeartbeatResource`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

vpapavas committed:
initial implementation of heartbeat and cluster stats=us

fixed tests

use application_server config to determine local host address

fixed compile issues

added extra tests

test

fixed failing test

added debug logging, made critical section smaller

addressed almogs comments

added return

stevenpyzhang committed:
fix: UncaughtExceptionHandler not being set for Persistent Queries (#4087)

blueedgenick committed:
docs: intent for klip-16: Introduce K$ Dynamic Views (#4056)

vcrfxia committed:
chore: issue regular LIST TOPICS from healthcheck, not extended (#4061)

JimGalasyn committed:
docs: update example output from SHOW TOPICS (KSQL-2981) (#4089)

JimGalasyn committed:
Merge branch '5.4.x'

vcrfxia committed:
docs: fix docker image names + remove monitoring interceptor docs (#4076)

vcrfxia committed:
chore: update KSQL_PULL_QUERIES_ENABLE_CONFIG name (MINOR) (#4059)

purplefox committed:
fix: Explicitly disallow table functions with table sources, fixes #4033 (#4085)

big-andy-coates committed:
test: have QTT support checking full schema of sources (#4083)

With primitive keys coming the QTT test needs to be able to test the full schema is correct, not just the value schema.

JimGalasyn committed:
docs: fix munged code blocks in Clickstream tutorial (DOCS-3084) (#4102)

derekjn committed:
chore: update the status of approved KLIPs

JimGalasyn committed:
docs: note CREATE CONNECTOR works only in interactive mode (DOCS-3036) (#4103)

JimGalasyn committed:
Merge branch '5.4.x'

JimGalasyn committed:
docs: note CREATE CONNECTOR works only in interactive mode (md docs) (#4104)

big-andy-coates committed:
chore: primitive keys for simple queries (#4096)

First of a few commits to start introducing support for primitive keys in different query types.

This commit opens the door for CT/CS statements with primitive keys, (`STRING`, `INT`, `BIGINT`, `BOOLEAN` and `DOUBLE`), and for using those sources in non-join, non-aggregate and non-partition-by queries.

rodesai committed:
test: serialize/deserialize plans from qtt (#4080)

Extends qtt to build the query plan and deserialize/serialize it
before executing. Should be enough to make sure plans are serializable
and executable until we improve qtt.

agavra committed:
fix: properly set key when partition by ROWKEY and join on non-ROWKEY (#4090)

big-andy-coates committed:
chore: partition-by primitive key support (#4098)

* chore: partition-by primitive key support

Fixes: https://github.com/confluentinc/ksql/issues/4092

WIP: This commit gets `PARTITION BY` clauses working with primitive key types. However, it does disable a couple of join until https://github.com/confluentinc/ksql/issues/4094 has been completed.

BREAKING CHANGE: A `PARTITION BY` now changes the SQL type of `ROWKEY` in the output schema of a query.

For example, consider:

```sql
CREATE STREAM INPUT (ROWKEY STRING KEY, ID INT) WITH (...);
CREATE STREAM OUTPUT AS SELECT ROWKEY AS NAME FROM INPUT PARTITION BY ID;
```

Previously, the above would have resulted in an output schema of `ROWKEY STRING KEY, NAME STRING`, where `ROWKEY` would have stored the string representation of the integer from the `ID` column.  With this commit the output schema will be `ROWKEY INT KEY, NAME STRING`.

apurvam committed:
Add milestone 0.7.0 to all new bugs.

Adding the next release milestone (0.7.0 at this time) to every bug report so that they are triaged weekly and can be prioritized accordingly.

apurvam committed:
Changing the default milestone to the milestone id

apurvam committed:
reverting the milestone field

the milestone field seems to have no effect, reverting.

rodesai committed:
refactor: nicer spec for aggregation steps (#4063)

* refactor: nicer spec for aggregation steps

This patch cleans up the aggregation step to specify a list of
non-aggregate column references instead of a count of non-aggregate
columns.

JimGalasyn committed:
docs: fix markdown docs build (DOCS-3091) (#4112)

agavra committed:
chore: split ARRAYCONTAINS into JSON_ARRAY_CONTAINS and ARRAY_CONTAINS (#4105)

* chore: split ARRAYCONTAINS into JSON_ARRAY_CONTAINS and ARRAY_CONTAINS

BREAKING CHANGE: the ARRAYCONTAINS function now needs to be referenced
as either JSON_ARRAY_CONTAINS or ARRAY_CONTAINS depending on the
intended param types

agavra committed:
chore: remove RUN SCRIPT from the server implementation (#4116)

BREAKING CHANGE: commands that were persisted with RUN SCRIPT will no
longer be executable

agavra committed:
chore: the KsqlJsonDeserializer no longer guarantees ordering (#4115)

stevenpyzhang committed:
fix: NPE when starting StandaloneExecutor (#4119)

big-andy-coates committed:
chore: group-by primitive key support (#4108)

* chore: group-by primitive key support

Fixes: https://github.com/confluentinc/ksql/issues/4092

This commit gets `GROUP BY` clauses working with primitive key types.

BREAKING CHANGE: A `GROUP BY` on single expressions now changes the SQL type of `ROWKEY` in the output schema of the query to match the SQL type of the expression.

 For example, consider:

 ```sql
 CREATE STREAM INPUT (ROWKEY STRING KEY, ID INT) WITH (...);
 CREATE TABLE OUTPUT AS SELECT COUNT(*) AS COUNT FROM INPUT GROUP BY ID;
 ```

 Previously, the above would have resulted in an output schema of `ROWKEY STRING KEY, COUNT BIGINT`, where `ROWKEY` would have stored the string representation of the integer from the `ID` column.

 With this commit the output schema will be `ROWKEY INT KEY COUNT BIGINT`.

BREAKING CHANGE: Any`GROUP BY` expression that resolves to `NULL`, including because a UDF throws an exception, now results in the row being excluded from the result.  Previously, as the key was a `STRING` a value of `"null"` could be used. With other primitive types this is not possible. As key columns must be non-null any exception is logged and the row is excluded.

big-andy-coates committed:
chore: reserve KLIP numbers

big-andy-coates committed:
chore: reserve KLIP numbers

bbejeck committed:
chore: Update KafkaStreams.close to take a duration parameter (#4110)

Updated QueryMetadata#close method to use KafkaStreams#close(Duration.ofXXX) vs. the no-arg version. When calling KafkaStreams#close() the default timeout is set to Long.MAX_VALUE so the close call will block forever.

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Andy Coates <andy@confluent.io>,
 Almog Gavra <almog@confluent.io>

big-andy-coates committed:
chore: reserve klip number and update some links

big-andy-coates committed:
Merge branch 'master' of github.com:confluentinc/ksql

rodesai committed:
feat: support writing plans to command topic (#4106)

This patch adds support for writing the physical plan to the command
topic in place of the original KSQL statements.

Part of this change is a small refactor to our rest-app to engine communication
module (ksql.rest.server.computation):
- CommandQueue.enqueue now accepts CommandId and Command as parameters rather
  than computing them.
- DistributingExecutor no longer takes a RequestValidator. Instead, it accepts
  a factory for creating validated commands. A validated command is one that has
  been validated to execute successfully against an engine snapshot.
  RequestValidator also uses an instance of this factory to validate statements
  at the api endpoint layer.

For now, writing of query plans is enabled by a config (ksql.execution.plan.enable)
which defaults to false. We'll drop this config once all the plan schema changes
are merged.

ConfluentJenkins committed:
Set Confluent to 5.3.2, Kafka to 5.3.2.

alex-dukhno committed:
feat: implemention of KLIP-13 (#4099)

alex-dukhno committed:
feat: show properties now includes embedded connect properties and scope (#4099)

JimGalasyn committed:
docs: update push and pull query topics with feedback (DOCS-3092) (#4117)

* docs: update push and pull query topics with feedback (DOCS-3092)

* docs: incorporate feedeback

JimGalasyn committed:
docs: fix a typo in queries overview topic (DOCS-3105) (#4133)

JimGalasyn committed:
docs: add functions index to markdown docs (DOCS-3049) (#4135)

rodesai committed:
chore: fix join left right names (#4136)

* chore: rename join left/right to leftSource/rightSource

* chore: fix schema

albertosantini committed:
fix: untracked file after cloning on Windows (#4122)

* fix: untracked file after cloning on Windows

This patch resolves a problem on Windows.
After cloning the repo, git status shows one untracked file.

This is a due to the trailing period in the file name.

stevenpyzhang committed:
feat: add source statement to SourceDescription (#4134)

agavra committed:
feat: add COUNT_DISTINCT and allow generics in UDAFs (#4150)

big-andy-coates committed:
Adds support for using primitive types in joins. (#4132)

* chore: primitive key support in JOINs

Adds support for using primitive types in joins.

BREAKING CHANGE: Some existing joins may now fail and the type of `ROWKEY` in the result schema of joins may have changed.

When `ROWKEY` was always a `STRING` it was possible to join an `INTEGER` column with a `BIGINT` column.  This is no longer the case. A `JOIN` requires the join columns to be of the same type. (See https://github.com/confluentinc/ksql/issues/4130 which tracks adding support for being able to `CAST` join criteria).

Where joining on two `INT` columns would previously have resulted in a schema containing `ROWKEY STRING KEY`, it would not result in `ROWKEY INT KEY`.

stevenpyzhang committed:
feat: add config to make error messages configurable (#4121)

* feat: add config to make error messages configurable

* refactor the config

* changes

JimGalasyn committed:
docs: update codeowners file for docs-md directory (DOCS-3120) (#4161)

stevenpyzhang committed:
fix: show topics doesn't display topics with different casing (#4159)

vcrfxia committed:
test: do not mock config store in StandaloneExecutor integration test (#4128)

vcrfxia committed:
Merge branch '5.4.x'

agavra committed:
fix: decimals in structs should display as numeric (#4165)

rayokota committed:
test: disambiguate some Schema Registry methods (#4166)

As part of adding support to Schema Registry for additional schema
types beyond Avro, some methods on SchemaRegistryClient will be
overloaded.  This change is to disambiguate these method calls in
the KSQL tests.

big-andy-coates committed:
chore: simplify group by schema resolving code (#4154)

As per comment here: https://github.com/confluentinc/ksql/pull/4108/files#r357023751

vinothchandar commented on ksql-execution/src/main/java/io/confluent/ksql/services/SimpleKsqlClient.java:
```diff
@@ -34,4 +37,14 @@
       String sql
   );
+
+  RestResponse<HeartbeatResponse> sendHeartbeatRequest(
```
name them `makeHeartbeatRequest` like the other methods for consistency?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+    return new HeartbeatHandler.Builder();
+  }
+
+  private HeartbeatHandler(KsqlEngine engine, ServiceContext serviceContext, int threadPoolSize,
```
should this be formatted like the other methods.. one argument per line? there are may be other places in the file as well

vinothchandar replied:
Can we pull the heartbeat related configs into its own class, so it's easier to read.

vpapavas replied:
We don't enforce the checkstyle of one argument per line anymore (at least that's where I was left off in that discussion)

vinothchandar commented on ksql-execution/src/main/java/io/confluent/ksql/services/SimpleKsqlClient.java:
```diff
@@ -34,4 +37,14 @@
+      long timestamp
+  );
+
+  RestResponse<ClusterStatusResponse> getClusterStatusRequest(
```
same here

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+  }
+
+  private HeartbeatHandler(KsqlEngine engine, ServiceContext serviceContext, int threadPoolSize,
+                           long heartbeatSendInterval, long heartbeatCheckInterval,
```
nit: `IntervalMs` to denote the time unit that these intervals are in?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+  }
+
+  void buildLocalHostInfo(final URI localServerURL) {
+
```
nit:extra line

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+    receivedHeartbeats.clear();
+  }
+
+  void buildLocalHostInfo(final URI localServerURL) {
```
`localServerURI`?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+    }
+  }
+
+  class ProcessHeartbeat implements Runnable {
```
name these classes ending with `Task`?  `ProcessHeartbeatTask` and so on?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+    public void run() {
+      try {
+        //broadcast heartbeat
+        remoteHostsURI.forEach((uri, bool) -> serviceContext.getKsqlClient().sendHeartbeatRequest(
```
we should probably move the exception handling within the forEach. Otherwise, even if one host errors out, it will abort sending heartbeats to everyone.

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+        remoteHostsURI.forEach((uri, bool) -> serviceContext.getKsqlClient().sendHeartbeatRequest(
+            uri, localHostInfo, System.currentTimeMillis()));
+      } catch (Exception e) {
+        e.printStackTrace();
```
can we log this instead?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+          .map(queryMetadata -> ((QueryMetadata)queryMetadata).getKafkaStreams().allMetadata())
+          .flatMap(Collection::stream)
+          .map(StreamsMetadata::hostInfo)
+          .filter(hostInfo -> !(hostInfo.host().equals(localHostInfo.host())
```
general comment.. we should make sure the advertised hostname (one in streams metadata), matches with the `localHostInfo`. Just discuss this once with alan, as the advertised listeners are being reworked now

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+
+      for (HostInfo hostInfo : uniqueHosts) {
+        // First time discovered
+        if (remoteHostsInfo.putIfAbsent(hostInfo, true) == null) {
```
what does `true` signify here? I don't think we ever set it to false? can it just be a `Set`, if so?

vpapavas replied:
There is no concurrent set implementation, that's why I make it a Map.

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+
+  public static class Builder {
+
+    private int nestedThreadPoolSize;
```
why the `nested` prefix? can we have time units for all these configs?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java:
```diff
@@ -0,0 +1,341 @@
+      for (Entry<HostInfo, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        //Question: Until this thread gets the lock, "now" will have changed. But if I compute "now"
```
discussed.. may be better to leave a GH comment for this in the future?  so you dont have have change code to remove the question? :)

vpapavas replied:
You mean when I do the actual PR, to put this question as a comment ?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -285,6 +293,11 @@ private void initialize() {
     );
 
+    heartbeatHandler.buildLocalHostInfo(ServerUtil.getServerAddress(restConfig));
+    heartbeatHandler.scheduleHeartbeatSender();
```
can we move all these 3 schedules into a method inside  heartbeatHandler?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -285,6 +293,11 @@ private void initialize() {
         serviceContext
     );
 
+    heartbeatHandler.buildLocalHostInfo(ServerUtil.getServerAddress(restConfig));
```
any reason this cannot be passed in during the `heartbeatHandler` construction?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java:
```diff
@@ -77,6 +77,36 @@
+  private static final String KSQL_HEARTBEAT_SEND_INTERVAL_MS_DOC =
+      "Interval at which heartbeats are broadcasted to servers.";
+
+  public static final String KSQL_HEARTBEAT_CHECK_INTERVAL_MS_CONFIG =
```
check is a better name. also rename `ProcessHeartBeatTask` to `CheckHeartBeatTask`?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java:
```diff
@@ -77,6 +77,36 @@
+  public static final String KSQL_HEARTBEAT_WINDOW_MS_CONFIG =
+      KSQL_CONFIG_PREFIX + "heartbeat.window.ms";
+  private static final String KSQL_HEARTBEAT_WINDOW_MS_DOC =
+      "The size of the window for which to count missed heartbeat.";
```
hearbeat**s**?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java:
```diff
@@ -77,6 +77,36 @@
+  public static final String KSQL_HEARTBEAT_MISSED_THRESHOLD_CONFIG =
+      KSQL_CONFIG_PREFIX + "heartbeat.missed.threshold.ms";
+  private static final String KSQL_HEARTBEAT_MISSED_THRESHOLD_DOC =
+      "Minimum number of consecutive missed heartbeats that constitute a server as down.";
```
flag a server as down?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java:
```diff
@@ -77,6 +77,36 @@
+  public static final String KSQL_HEARTBEAT_DISCOVER_CLUSTER_MS_CONFIG =
+      KSQL_CONFIG_PREFIX + "heartbeat.discover.interval.ms";
+  private static final String KSQL_HEARTBEAT_DISCOVER_CLUSTER_MS_DOC =
+      "Interval at which server discovers what remote servers exist in cluster.";
```
attempts to discover what other ksql servers exist in the cluster?

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/HeartbeatResource.java:
```diff
@@ -0,0 +1,63 @@
+    TreeMap<Long, HeartbeatInfo> heartbeats = heartbeatHandler.getReceivedHeartbeats()
+        .get(hostInfo);
+
+    //Question: is this correct?
```
remove?

vinothchandar commented on ksql-rest-model/src/main/java/io/confluent/ksql/rest/entity/HostInfoEntity.java:
```diff
@@ -0,0 +1,70 @@
+  private final int port;
+
+  @JsonCreator
+  public HostInfoEntity(
```
do we test the other entities in the current code for ser/deser? if so, then we should also add some tests

vpapavas replied:
I haven't found any tests

vinothchandar commented on ksql-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTarget.java:
```diff
@@ -74,6 +79,20 @@ public KsqlTarget authorizationHeader(final String authHeader) {
+  }
+
+  public RestResponse<ClusterStatusResponse> getClusterStatus() {
+    return get("/clusterStatus", ClusterStatusResponse.class);
```
pull this to a variable as well?

vinothchandar commented on ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/TestKsqlRestApp.java:
```diff
@@ -401,6 +402,9 @@ private static KsqlRestConfig buildConfig(
     configMap.put(KsqlConfig.KSQL_STREAMS_PREFIX + "cache.max.bytes.buffering", 0);
     configMap.put(KsqlConfig.KSQL_STREAMS_PREFIX + "auto.offset.reset", "earliest");
     configMap.put(KsqlConfig.KSQL_ENABLE_UDFS, false);
+    configMap.put(KsqlRestConfig.KSQL_HEARTBEAT_SEND_INTERVAL_MS_CONFIG, 600000);
```
this is so that it does not heartbeat by default?

vpapavas replied:
Yes, else the tests fail as they have a disabled ksql client that is not allowed to do requests

vinothchandar reviewed (commented):
I still have to review HeartbeatHandler and the two tests more closely.. Will resume..

big-andy-coates committed:
chore: enforce WITH KEY column type matches ROWKEY type (#4147)

* chore: enforce WITH KEY column type matches ROWKEY type

BREAKING CHANGE:  Any `KEY` column identified in the `WITH` clause must be of the same Sql type as `ROWKEY`.

Users can provide the name of a value column that matches the key column, e.g.

```sql
CREATE STREAM S (ID INT, NAME STRING) WITH (KEY='ID', ...);
```

Before primitive keys was introduced all keys were treated as `STRING`. With primitive keys `ROWKEY` can be types other than `STRING`, e.g. `BIGINT`.
It therefore follows that any `KEY` column identified in the `WITH` clause must have the same SQL type as the _actual_ key,  i.e. `ROWKEY`.

With this change the above example statement will fail with the error:

```
The KEY field (ID) identified in the WITH clause is of a different type to the actual key column.
Either change the type of the KEY field to match ROWKEY, or explicitly set ROWKEY to the type of the KEY field by adding 'ROWKEY INTEGER KEY' in the schema.
KEY field type: INTEGER
ROWKEY type: STRING
```

As the error message says, the error can be resolved by changing the statement to:

```sql
CREATE STREAM S (ROWKEY INT KEY, ID INT, NAME STRING) WITH (KEY='ID', ...);
```

agavra committed:
feat: add support for inline struct creation (#4120)

rodesai committed:
fix: change query id generation to work with planned commands (#4149)

* fix: change query id generation to work with planned commands

This patch changes up how we generate query IDs to play nice with
planned commands. Before this change, statements would get the current
offset as their query id. However planned commands get their query IDs
before being enqueued, so they should really get the _next_ expected
offset as their ID. This patch changes up the id generation to work
this way. The next ID is set _after_ statemetns/plans are executed,
and is set to the next expected offset.

Co-Authored-By: Victoria Xia <victoria.f.xia281@gmail.com>

ConfluentJenkins committed:
Merge branch '5.3.2-post' into 5.3.x

ConfluentJenkins committed:
Merge branch '5.3.x' into 5.4.x

ConfluentJenkins committed:
Merge branch '5.4.x'

big-andy-coates committed:
fix: pull queries should work across nodes (#4169)

* fix: pull queries should work across nodes

Fixes: https://github.com/confluentinc/ksql/issues/4142
Fixes: https://github.com/confluentinc/ksql/issues/4151
Fixes: https://github.com/confluentinc/ksql/issues/4152

Introduces a new `inter.node.listener` that can be used to specify a URL that the node can be contacted on by other nodes.  This can be different to the listeners defined in `listeners`. This can be required if `listeners` is set to a wildcard address, i.e. IPv4 `0.0.0.0` or IPv6 `[::]`, or if the node sits behind network infrastructure that requires other nodes to reach it using a different URL.

If `inter.node.listener` is not set it still defaults to the first listener in `listener` config. However, it now replaces an wildcard address with `localHost`. This means inter-node comms is still possible for nodes running on the same host.

Warnings are logged if the inter-node listener resolves to a loopback or local address.

big-andy-coates committed:
docs: add basic docker-compose file and instructions (#4175)

* docs: add basic docker-compose file and instructions

agavra committed:
fix: immutability in some more classes (MINOR) (#4179)

ConfluentJenkins committed:
Bump Confluent to 5.3.3-SNAPSHOT, Kafka to 5.3.3-SNAPSHOT

ConfluentJenkins committed:
Merge branch '5.3.x' into 5.4.x

ConfluentJenkins committed:
Merge branch '5.4.x'

big-andy-coates committed:
chore: pull query support for primitive keys (#4178)

rodesai committed:
test: qtt generate/validate execution plans (#4176)

* feat: qtt generate/validate execution plans

This patch extends our query validation tests to generate, execute, and
validate execution plans.

TestExecutor/TestExecutorUtil have been extended to support executing from
a plan as an alternative to a statement, if included in the TestCase spec.
In both cases, TestExecutorUtil first builds an Iterable that yields
instances of ConfiguredKsqlPlan. If the plans are included in the TestCase,
the Iterable just iterates over the supplied plans. If not, the Iterable
builds each statement into a plan.

Execution plans are derived from the qtt test cases and checked into the
repository under `ksql-functional-tests/src/test/resources/historical_plans`.
The plans are organized as follows:
`<test-case>`: all the plans for a given test case
`<test-case>/<version>_<timestamp>`: the plan for a given test case at a
                                     point in time
`<test-case>/<version>_<timestamp>`/spec.json - a json containing the config,
                                                plan, and any generated schemas.
`<test-case>/<version>_<timestamp>`/topology - the streams topology description.

A new test loader called PlannedTestLoader iterates over all the plans and
builds instances of TestCase out of them.

To ensure we include new query plans over time, this patch includes a test
called PlannedTestUpToDateTest, that iterates over each qtt test case and
ensures that we have persisted the current plan that the engine builds.

Finally, this patch includes PlannedTestGenerator which you can use to
generate any new plans that need to be generated.

agavra committed:
fix: reintroduce FetchFieldFromStruct as a public UDF (#4185)

agavra committed:
Merge branch '5.4.x'

stevenpyzhang committed:
chore: use UsePartitionTimeOnInvalidTimestamp instead of UsePreviousTimeOnInvalidTimeStamp (#4199)

big-andy-coates committed:
chore: re-enable final param and var checkstyle rule

big-andy-coates committed:
test: enhance test framework to support primitive keys (#4201)

* test: enhance test framework to support primitive keys

Enhanced the test framework to work with test cases with `BIGINT` and `DOUBLE` keys. This was previously failing due to Jackson deserializing the JSON describing the expected values into to `INT` and `DECIMAL`.

To achieve this I've started to tidy up the mess in the test framework. Specifically, we no longer track the serde to use in the `Topic` class, as this is constructed when parsing the test cases. Doing so was problematic as it was hard to know what serde should be used just from the test case text.  The serde to use is now determined by looking at the format of sources in the metastore once the SQL statements have been executed.

Also includes some additional test cases for pull queries working with primitive keys, and pulling out the code that determines the format / serde to use for a topic into a `TopicInfoCache`, as this is not shared by `QTT` and `RQTT` tests.

JimGalasyn committed:
docs: unpin mkdocs-macros-plugin version (DOCS-3100) (#4208)

JimGalasyn committed:
docs: add security note linking to processing log settings (DOCS-3076) (#4211)

JimGalasyn committed:
Merge branch '5.4.x'

JimGalasyn committed:
docs: add note to ksqlDB processing log settings for security (#4213)

JimGalasyn committed:
docs: add section for ksql.streams.state.dir to md docs (#4214)

JimGalasyn committed:
docs: add section for ksql.streams.state.dir (DOCS-3153) (#4210)

* docs: add section for ksql.streams.state.dir (DOCS-3153)

* docs: add note about uniqueness of state store dir

* docs: update default state store dir per feedback

JimGalasyn committed:
Merge branch '5.4.x'

big-andy-coates committed:
chore: drop square brackets from string representation of logical schema (MINOR) (#4203)

* chore: drop square brackets from string representation of logical schema

They are not needed and stop the string representation being easily used, e.g. to build the schema in a CS/CT statement. Invariably, code has to strip or replace the square brackets.

big-andy-coates committed:
chore: fix TestDataProvider's to have correct key type (#4206)

* chore: fix TestDataProvider's to have correct key type

A couple of the `TestDataProvider` implementations have a `WITH(KEY)` field that has a non-`STRING` SQL type, but the SQL type of `ROWKEY` was still `STRING`.
This is not incorrect: the with-key field must have the same SQL type as the actual key, i.e. `ROWKEY`.

This change fixes the `TestDataProvider`s and any associated tests.  It also refactors the providers and the code tests use to produce and consume messages to clean this up and remove duplicate functionality.

JimGalasyn committed:
docs: remove basic auth section from server-config (DOCS-3140) (#4209)

JimGalasyn committed:
Merge branch '5.4.x'

purplefox committed:
refactor: Use FunctionName not String in FunctionRegistry API (#4225)

rodesai committed:
chore: add back expected topology tests (#4207)

* chore: add back expected topology tests

Temporarily add back expected topology tests, until we make the
full switchover to the new query plan based tests

rodesai committed:
test: add equals testers for plan classes (#4189)

test: add equals testers for plan classes

elismaga committed:
Disable building docker images (#4230)

elismaga committed:
Disable building docker images (#4230)

big-andy-coates committed:
chore: fix result schema on table GROUP BY for primitive keys (MINOR) (#4205)

* chore: fix result schema on table GROUP BY for primitive keys

The intermediate schema was incorrect as it wasn't updating the SQL type of ROWKEY to match the grouping column.

Also, added test to `KeyField` to enforce the key field matches the type of the key column.

big-andy-coates committed:
Merge branch '5.4.x'

spena committed:
refactor: inject a KsqlSecurityContext to REST requests (#4184)

vcrfxia committed:
fix: CLI commands may be terminated with semicolon+whitespace (MINOR) (#4234)

rodesai committed:
feat: add test topology rewriter (#4204)

* feat: add test topology rewriter

This patch adds a rewriter tool for rewriting test topologies
in the event that the streams topology changes in a backwards
compatible way.

spena committed:
feat: add a KSQL cache for Kafka authorization validator (#4186)

Implement a cache for permission checks done by the KsqlAuthorizationValidator.

big-andy-coates committed:
test: fix build issue caused by new register method on SR client (#4247)

rayokota committed:
test: Fix Schema Registry tests after new SR API changes (#4243)

big-andy-coates committed:
chore: ensure sensible error on use of unsupported primitive key type (#4246)

fixes: https://github.com/confluentinc/ksql/issues/4113

Added QTT tests to ensure sensible error messages are output from `CT`/`CS` and `CTAS`/`CSAS` statements where the key format, (currently always `KAFKA`), does not support the Sql type of the key column, e.g. the `KAFKA` format does not support `BOOLEAN`, `DECIMAL`, `MAP`, ARRAY` or `STRUCT` types.

Error message will look something like:

```
Key format does not support key schema.
format: KAFKA
schema: Persistence{schema=STRUCT<ROWKEY MAP<VARCHAR, INT>> NOT NULL, unwrapped=false}
reason: The 'KAFKA' format does not support type 'MAP'
```

rodesai committed:
fix: don't load current qtt test case from legacy loader (#4245)

big-andy-coates committed:
chore: remove upstream docker registry as not public

AlanConfluent committed:
perf: Improves pull query performance by making the default schema service a singleton (#4216)

* perf: Improves pull query performance by making the schema registry client a singleton.

spena committed:
feat: enable Kafla ACL authorization checks for Pull Queries (#4187)

big-andy-coates committed:
fix: better error message on self-join (#4248)

Fixes: https://github.com/confluentinc/ksql/issues/4241

Self-joins are not yet supported. Previously they resulted in an confusing error message:

>  Invalid topology: Topic <something> has already been registered by another source.

They not result in:

> Can not join 'something' to 'something': self joins are not yet supported.

big-andy-coates committed:
fix: include path of field that causes JSON deserialization error (#4249)

Fixes: https://github.com/confluentinc/ksql/issues/4238

For complex types, the current deserialization error can be a bit lacking, as it does not include details of which field causes the error.  This commit enhances the JSON deserializer to include the JSON path of where the error occurred in the error message.

The design keep the happy path quick by using the call stack to maintain the stack of fields, rather than tracking them in some separate stack in memory. Figured this was worth while for such a performance sensitive area of the code.  There's no detectable perf hit from this change.

I've also increased the times on the `SerdeBenchmark`, as from my own experience 10 seconds is not enough time for the JVM to optimise the byte code.

Perf test on JSON deserialization before changes:

```
# JMH version: 1.21
# VM version: JDK 1.8.0_162, Java HotSpot(TM) 64-Bit Server VM, 25.162-b12
# VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_162.jdk/Contents/Home/jre/bin/java
# VM options: -javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=58499:/Applications/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
# Warmup: 6 iterations, 30 s each
# Measurement: 3 iterations, 60 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: io.confluent.ksql.benchmark.SerdeBenchmark.deserialize
# Parameters: (schemaName = metrics, serializationFormat = JSON)

# Run progress: 0.00% complete, ETA 00:06:00
# Fork: 1 of 1
# Warmup Iteration   1: 5.188 us/op
# Warmup Iteration   2: 5.124 us/op
# Warmup Iteration   3: 5.193 us/op
# Warmup Iteration   4: 5.529 us/op
# Warmup Iteration   5: 5.378 us/op
# Warmup Iteration   6: 5.266 us/op
Iteration   1: 5.209 us/op
Iteration   2: 5.183 us/op
Iteration   3: 5.274 us/op

Result "io.confluent.ksql.benchmark.SerdeBenchmark.deserialize":
  5.222 ±(99.9%) 0.854 us/op [Average]
  (min, avg, max) = (5.183, 5.222, 5.274), stdev = 0.047
  CI (99.9%): [4.368, 6.075] (assumes normal distribution)

# Run complete. Total time: 00:06:01

REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.

Benchmark                   (schemaName)  (serializationFormat)  Mode  Cnt  Score   Error  Units
SerdeBenchmark.deserialize       metrics                   JSON  avgt    3  5.222 ± 0.854  us/op
 ```

After:

```
# JMH version: 1.21
# VM version: JDK 1.8.0_162, Java HotSpot(TM) 64-Bit Server VM, 25.162-b12
# VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_162.jdk/Contents/Home/jre/bin/java
# VM options: -javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=60943:/Applications/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
# Warmup: 6 iterations, 30 s each
# Measurement: 3 iterations, 60 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: io.confluent.ksql.benchmark.SerdeBenchmark.deserialize
# Parameters: (schemaName = metrics, serializationFormat = JSON)

# Run progress: 0.00% complete, ETA 00:06:00
# Fork: 1 of 1
# Warmup Iteration   1: 5.284 us/op
# Warmup Iteration   2: 5.114 us/op
# Warmup Iteration   3: 5.185 us/op
# Warmup Iteration   4: 5.105 us/op
# Warmup Iteration   5: 5.171 us/op
# Warmup Iteration   6: 5.119 us/op
Iteration   1: 5.208 us/op
Iteration   2: 5.248 us/op
Iteration   3: 5.198 us/op

Result "io.confluent.ksql.benchmark.SerdeBenchmark.deserialize":
  5.218 ±(99.9%) 0.477 us/op [Average]
  (min, avg, max) = (5.198, 5.218, 5.248), stdev = 0.026
  CI (99.9%): [4.740, 5.695] (assumes normal distribution)

# Run complete. Total time: 00:06:02

REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.

Benchmark                   (schemaName)  (serializationFormat)  Mode  Cnt  Score   Error  Units
SerdeBenchmark.deserialize       metrics                   JSON  avgt    3  5.218 ± 0.477  us/op

```

agavra committed:
feat: allow environment variables to configure embedded connect (#4260)

albertosantini committed:
chore: remove unused import (#4273)

ConfluentJenkins committed:
Set Confluent to 5.4.0, Kafka to 5.4.0.

purplefox committed:
docs: klip-15 New client and API (#4069)

big-andy-coates committed:
fix: add logging during restore (#4270)

Fixes: https://github.com/confluentinc/ksql/issues/4269

spena committed:
feat: ask for password if -p is not provided (#4153)

big-andy-coates committed:
fix: report clearer error message when AVG used with DELIMITED (#4295)

See https://github.com/confluentinc/ksql/issues/4294

`AVG` doesn't work with `DELIMITED` format and the error message isn't great.

Example statements that cause the error:

```sql
-- Given:
CREATE STREAM INPUT (VALUE integer) WITH (kafka_topic='test_topic', value_format='DELIMITED');

-- When:
CREATE TABLE OUTPUT AS SELECT avg(value) AS avg FROM INPUT group by ROWKEY;
```

Old error message:

```
ksql> CREATE TABLE OUTPUT AS SELECT avg(value) AS avg FROM INPUT group by ROWKEY;
CREATE TABLE OUTPUT AS SELECT avg(value)Value format does not support value schema.
format: DELIMITED
schema: Persistence{schema=STRUCT<KSQL_INTERNAL_COL_0 INT, KSQL_INTERNAL_COL_1 VARCHAR, KSQL_AGG_VARIABLE_0 STRUCT<SUM INT, COUNT BIGINT>> NOT NULL, unwrapped=false}
reason: The 'DELIMITED' format does not support type 'STRUCT'
Caused by: The 'DELIMITED' format does not support type 'STRUCT'
```

This PR improves the error message a bit:

New error message:

```
One of the functions used in the statement has an intermediate type that the value format can not handle. Please remove the function or change the format.
Consider up-voting https://github.com/confluentinc/ksql/issues/3950, which will resolve this limitation
Caused by: Value format does not support value schema.
format: DELIMITED
schema:
	Persistence{schema=STRUCT<KSQL_INTERNAL_COL_0 INT, KSQL_INTERNAL_COL_1 VARCHAR,
	KSQL_AGG_VARIABLE_0 STRUCT<SUM INT, COUNT BIGINT>> NOT NULL,
	unwrapped=false}
reason: The 'DELIMITED' format does not support type 'STRUCT'
Caused by: The 'DELIMITED' format does not support type 'STRUCT'
```

elismaga committed:
build: Do not run twist lock scan and other docker image operations since we no longer build images in this repo (#4299)

agavra committed:
Merge branch '5.4.x'

agavra committed:
chore: migrate to use ParsedSchema for SchemaRegistryClient (#4259)

derekjn committed:
docs: add embedded Connect tutorial (#4212)

vpapavas committed:
docs: KLIP 12 - Implement High-Availability for Pull queries (#4022)

* Adding doc for klip-12

Co-authored-by: vinoth chandar <vinothchandar@users.noreply.github.com>

big-andy-coates committed:
fix: deadlock when closing transient push query (#4297)

* fix: deadlock when closing transient push query

fixes: https://github.com/confluentinc/ksql/issues/4296

The produce side not calls `offer` in a loop, with a short timeout, to try and put the row into the blocking queue. When the consume side closes the query, e.g. on an `EOFException` if the user has closed the connection, the query first closes the queue; setting a flag the producers are checking on each loop; causing any producers to exit the loop. Then it can safely close the KS topology.

big-andy-coates committed:
fix: deadlock when closing transient push query (#4297)

fixes: https://github.com/confluentinc/ksql/issues/4296

The produce side not calls `offer` in a loop, with a short timeout, to try and put the row into the blocking queue. When the consume side closes the query, e.g. on an `EOFException` if the user has closed the connection, the query first closes the queue; setting a flag the producers are checking on each loop; causing any producers to exit the loop. Then it can safely close the KS topology.

(cherry picked from commit 6b5ce0c1a6f34891dcac88ad7454a06f3ed08f37)

big-andy-coates committed:
fix: log4j properties files (#4293)

* fix: log4j properties files

This PR removes the log4j.properties files from our production jars. The presence of such files will make it more tricky for clients to correctly configure log4j with their own properties files as log4j picks the first it finds on the class path by default.

Also, changed many log levels from `ERROR` to `WARN` as important warnings are being excluded, which could be hiding issues.

AlanConfluent committed:
perf: Avoids logging INFO for rest-util requests, since it hurts pull query performance (#4302)

* perf: Avoids logging INFO for rest-util requests, since it hurts pull query performance

big-andy-coates committed:
fix: fix NPE in CLI if not username supplied (#4312)

Fixes NPE when running the CLI with no arguments. Was throwing NPE on line:

```java
if (!options.getUserName().isEmpty() && !options.isPasswordSet()) {
```

in Ksql.java, as `options.getUserName()` was returning `null`.

Refactored so that username and password are never null, only empty.

Also refactored so that server will default to `http://localhost:8088`, even if other params are supplied.  Previously, it only defaulted to this if no params supplied.

big-andy-coates committed:
test: rQTT should fail if expecting more responses than statements (#4309)

This just improves the error message. Better this than an `IndexOutOfBoundsException`.

JimGalasyn committed:
docs: refactor installation docs around Docker (DOCS-3009) (#4235)

* docs: start refactoring installation docs (DOCS-3009)

* docs: add docker command lines for starting ksqldb-cli

* docs: fully dockerize the main installation topic

* docs: refactor around Docker stacks

* docs: rework installation topic around one stack file

* docs: add section for PostgreSQL stack

* docs: add Compose output

* docs: incorporate copy edits

big-andy-coates committed:
Revert "fix: deadlock when closing transient push query (#4297)"

This reverts commit ac8fb630

agavra committed:
docs: add docs for COUNT_DISTINCT (#4300)

rodesai committed:
Merge branch '5.4.x'

ConfluentJenkins committed:
Merge branch '5.4.0-post' into 5.4.x

ConfluentJenkins committed:
Merge branch '5.4.x'

agavra committed:
feat: expression support in JOINs (#4278)

ConfluentJenkins committed:
Bump Confluent to 5.4.1-SNAPSHOT, Kafka to 5.4.1-SNAPSHOT

ConfluentJenkins committed:
Merge branch '5.4.x'

big-andy-coates committed:
test: push queries work with non-windowed primitive keys (#4310)

* test: push queries work with non-windowed primitive keys

Fixes: https://github.com/confluentinc/ksql/issues/4123

big-andy-coates committed:
test: ensure test cases for group by and partition by primitive key (#4311)

Fixes: https://github.com/confluentinc/ksql/issues/4107

rmoff committed:
Add note re. logs folder for server output (#4318)

I followed the instructions to build and run locally, and fatal errors weren't written to stdout - took me a while to realise why the process was just exiting :) Found the error in the `ksql.log` log in the logs folder.

JimGalasyn committed:
docs: fix broken link on dev guide index page (DOCS-3243) (#4327)

rodesai committed:
fix: pin the jetty client version (#4324)

This patch pins the jetty client version to the jetty version specified
by confluent/common.

big-andy-coates committed:
test: added more table related test cases (#4329)

and fixed small bug in test framework

spena committed:
fix: add ksql-test-runner deps to ksql package lib (#4272)

big-andy-coates committed:
fix: report window type and query status better from API (#4313)

This commit:
1. exposes the window type of the key of a query/source, i.e. `HOPPING`, `TUMBLING` `SESSION` or none.
2. makes the status of a query easier to find.
3. fixes a bug that meant the statement text of a query was not displayed in the CLI.

BREAKING CHANGE: The response from the RESTful API has changed for some commands with this commit: the `SourceDescription` type no longer has a `format` field. Instead it has `keyFormat` and `valueFormat` fields.

## `SHOW QUERY` changes:

Response now includes a `state` property for each query that indicates the state of the query.

e.g.

```json
{
  "queryString" : "create table OUTPUT as select * from INPUT;",
  "sinks" : [ "OUTPUT" ],
  "id" : "CSAS_OUTPUT_0",
  "state" : "Running"
}
```

The CLI output was:

```
 ksql> show queries;

  Query ID                   | Kafka Topic         | Query String

   CSAS_OUTPUT_0              | OUTPUT              | CREATE STREAM OUTPUT WITH (KAFKA_TOPIC='OUTPUT', PARTITIONS=1, REPLICAS=1) AS SELECT *
 FROM INPUT INPUT
 EMIT CHANGES;
  CTAS_CLICK_USER_SESSIONS_5 | CLICK_USER_SESSIONS | CREATE TABLE CLICK_USER_SESSIONS WITH (KAFKA_TOPIC='CLICK_USER_SESSIONS', PARTITIONS=1, REPLICAS=1) AS SELECT
   CLICKSTREAM.USERID USERID,
   COUNT(*) COUNT
 FROM CLICKSTREAM CLICKSTREAM
 WINDOW SESSION ( 300 SECONDS )
 GROUP BY CLICKSTREAM.USERID
 EMIT CHANGES;

 For detailed information on a Query run: EXPLAIN <Query ID>;
```

and is now:

```
 Query ID                   | Status      | Kafka Topic         | Query String

 CSAS_OUTPUT_0              | RUNNING     | OUTPUT              | CREATE STREAM OUTPUT WITH (KAFKA_TOPIC='OUTPUT', PARTITIONS=1, REPLICAS=1) AS SELECT *FROM INPUT INPUTEMIT CHANGES;

For detailed information on a Query run: EXPLAIN <Query ID>;

```
Note the addition of the `Status` column and the fact that `Query String` is now longer being written across multiple lines.

## `DESCRIBE <source>;` changes:

old CLI output:

```
ksql> describe CLICK_USER_SESSIONS;

Name                 : CLICK_USER_SESSIONS
 Field   | Type

 ROWTIME | BIGINT           (system)
 ROWKEY  | INTEGER          (system)
 USERID  | INTEGER
 COUNT   | BIGINT

For runtime statistics and query details run: DESCRIBE EXTENDED <Stream,Table>;
```

New CLI output:

```
ksql> describe CLICK_USER_SESSIONS;

Name                 : CLICK_USER_SESSIONS
 Field   | Type

 ROWTIME | BIGINT           (system)
 ROWKEY  | INTEGER          (system) (Window type: SESSION)
 USERID  | INTEGER
 COUNT   | BIGINT

For runtime statistics and query details run: DESCRIBE EXTENDED <Stream,Table>;
```

Note the addition of the `Window Type` information.

The extended version of the command has also changed.

Old output:

```
ksql> describe extended CLICK_USER_SESSIONS;

Name                 : CLICK_USER_SESSIONS
Type                 : TABLE
Key field            : USERID
Key format           : STRING
Timestamp field      : Not set - using <ROWTIME>
Value Format                : JSON
Kafka topic          : CLICK_USER_SESSIONS (partitions: 1, replication: 1)
Statement            : CREATE TABLE CLICK_USER_SESSIONS WITH (KAFKA_TOPIC='CLICK_USER_SESSIONS', PARTITIONS=1, REPLICAS=1) AS SELECT
  CLICKSTREAM.USERID USERID,
  COUNT(*) COUNT
FROM CLICKSTREAM CLICKSTREAM
WINDOW SESSION ( 300 SECONDS )
GROUP BY CLICKSTREAM.USERID
EMIT CHANGES;

 Field   | Type

 ROWTIME | BIGINT           (system)
 ROWKEY  | INTEGER          (system)
 USERID  | INTEGER
 COUNT   | BIGINT

Queries that write from this TABLE
-----------------------------------
CTAS_CLICK_USER_SESSIONS_5 (RUNNING) : CREATE TABLE CLICK_USER_SESSIONS WITH (KAFKA_TOPIC='CLICK_USER_SESSIONS', PARTITIONS=1, REPLICAS=1) AS SELECT
 CLICKSTREAM.USERID USERID,
 COUNT(*) COUNT
 FROM CLICKSTREAM CLICKSTREAM
 WINDOW SESSION ( 300 SECONDS )
 GROUP BY CLICKSTREAM.USERID
 EMIT CHANGES;

For query topology and execution plan please run: EXPLAIN <QueryId>

Local runtime statistics
------------------------

(Statistics of the local KSQL server interaction with the Kafka topic CLICK_USER_SESSIONS)
```

New output:

```
ksql> describe extended CLICK_USER_SESSIONS;

Name                 : CLICK_USER_SESSIONS
Type                 : TABLE
Key field            : USERID
Timestamp field      : Not set - using <ROWTIME>
Key format           : KAFKA
Value format         : JSON
Kafka topic          : CLICK_USER_SESSIONS (partitions: 1, replication: 1)
Statement            : CREATE TABLE CLICK_USER_SESSIONS WITH (KAFKA_TOPIC='CLICK_USER_SESSIONS', PARTITIONS=1, REPLICAS=1) AS SELECT
  CLICKSTREAM.USERID USERID,
  COUNT(*) COUNT
FROM CLICKSTREAM CLICKSTREAM
WINDOW SESSION ( 300 SECONDS )
GROUP BY CLICKSTREAM.USERID
EMIT CHANGES;

 Field   | Type

 ROWTIME | BIGINT           (system)
 ROWKEY  | INTEGER          (system) (Window type: SESSION)
 USERID  | INTEGER
 COUNT   | BIGINT

Queries that write from this TABLE
-----------------------------------
CTAS_CLICK_USER_SESSIONS_5 (RUNNING) : CREATE TABLE CLICK_USER_SESSIONS WITH (KAFKA_TOPIC='CLICK_USER_SESSIONS', PARTITIONS=1, REPLICAS=1) AS SELECT  CLICKSTREAM.USERID USERID,  COUNT(*) COUNTFROM CLICKSTREAM CLICKSTREAMWINDOW SESSION ( 300 SECONDS ) GROUP BY CLICKSTREAM.USERIDEMIT CHANGES;

For query topology and execution plan please run: EXPLAIN <QueryId>

Local runtime statistics
------------------------

(Statistics of the local KSQL server interaction with the Kafka topic CLICK_USER_SESSIONS)
```

 Note: the change from `Key format` of `STRING` to `KAFKA`.  The output of `Window Type` information for windowed schemas and outputing sql statements on a single line.

agavra committed:
chore: change RUNNING to WARNING when all connector tasks fail (#4323)

rodesai committed:
chore: clean up LogicalSchemaWithMetaAndKeyFields (#4188)

vpapavas committed:
fix typo in comment
